File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> lastIndex = new HashMap<>();
        int n = s.length();
        int max = 0;
        int p1 = 0;
        for (int p2=0; p2<n; p2++)
        {
            char c = s.charAt(p2);
            Integer last = lastIndex.get(c);
            if (last!=null)
            {
                if (last>=p1)
                {
                    p1 = last+1;
                }
            }
            max = Math.max(max, p2-p1+1);
            lastIndex.put(c, p2);
        }
        return max;
        
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        last_indexes = {}
        max_len = 0
        start = -1
        for i, c in enumerate(s):
            if last_indexes.get(c, -1) > start:
                start = last_indexes[c]
            max_len = max(max_len, i-start)
            last_indexes[c] = i
        return max_len
        